// @framer/framer.base/code/Controls/Button.tsx
import * as React from "react"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { Button as BaseButton, ButtonProps, themes } from "../../../base"
import { ControlType, addPropertyControls } from "framer"
import {
    intentControls,
    decoratorControls,
    resizeControls,
    patternControls,
    themeControls,
} from "../propertyControls"
import { centerLayout, resizeLayout, OmitTheme } from "../utils"

type Props = OmitTheme<ButtonProps> &
    Partial<{
        theme: string
        text: string | any
        resize: string
        center: boolean
        withStartDecorator: boolean
        withEndDecorator: boolean
        iconStartDecorator: string
        iconEndDecorator: string
    }>

const defaults: Props = {
    width: 180,
    height: 50,
    theme: "light",
    text: "Hello world",
    intent: "primary",
    pattern: "primary",
    disabled: false,
    loading: false,
    resize: "none",
    center: true,
}

export function Button(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const {
        width,
        height,
        resize,
        text,
        withStartDecorator,
        iconStartDecorator,
        withEndDecorator,
        iconEndDecorator,
        center,
        theme,
        ...rest
    } = props
    const resizeDimensions = resizeLayout(resize, width, height)

    return centerLayout(
        props,
        <BaseButton
            endDecorator={withEndDecorator && iconEndDecorator}
            startDecorator={withStartDecorator && iconStartDecorator}
            {...rest}
            theme={themeContext || themes[theme]}
            {...resizeDimensions}
        >
            {text}
        </BaseButton>,
        center
    )
}

Button.defaultProps = {
    ...defaults,
}

addPropertyControls(Button, {
    ...themeControls(defaults.theme),
    ...intentControls(defaults.intent),
    ...patternControls(defaults.pattern),
    ...resizeControls(defaults.resize),
    disabled: {
        type: ControlType.Boolean,
        title: "Disabled",
        defaultValue: defaults.disabled,
    },
    loading: {
        type: ControlType.Boolean,
        title: "Loading",
        defaultValue: defaults.loading,
    },
    ...decoratorControls("ExternalLink"),
    text: {
        type: ControlType.String,
        title: "Text",
        defaultValue: defaults.text,
    },
})
