// @framer/framer.base/code/Controls/Input.tsx
import * as React from "react"
import { Input as BaseInput, InputProps, themes } from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { ControlType, addPropertyControls } from "framer"
import {
    intentControls,
    patternControls,
    themeControls,
    decoratorControls,
    resizeControls,
} from "../propertyControls"
import { centerLayout, resizeLayout, OmitTheme } from "../utils"

type Props = OmitTheme<InputProps> &
    Partial<{
        theme: string
        resize: string
        withStartDecorator: boolean
        withEndDecorator: boolean
        iconStartDecorator: boolean
        iconEndDecorator: boolean
    }>

const defaults: Props = {
    width: 240,
    height: 50,
    theme: "light",
    intent: "primary",
    pattern: "primary",
    type: "text",
    placeholder: "Hello world",
    value: "",
    disabled: false,
    resize: "none",
}

export function Input(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const {
        theme,
        resize,
        width,
        height,
        withStartDecorator,
        iconStartDecorator,
        withEndDecorator,
        iconEndDecorator,
        ...rest
    } = props
    const resizeDimensions = resizeLayout(resize, width, height)

    return centerLayout(
        props,
        <BaseInput
            endDecorator={withEndDecorator && iconEndDecorator}
            startDecorator={withStartDecorator && iconStartDecorator}
            {...rest}
            theme={themeContext || themes[theme]}
            {...resizeDimensions}
        />
    )
}

Input.defaultProps = {
    ...defaults,
}

addPropertyControls(Input, {
    ...themeControls(defaults.theme),
    ...intentControls(defaults.intent),
    ...patternControls(defaults.pattern),
    ...resizeControls(defaults.resize),
    type: {
        type: ControlType.Enum,
        title: "Type",
        defaultValue: defaults.type,
        options: [
            "date",
            "email",
            "number",
            "password",
            "search",
            "tel",
            "text",
            "time",
            "url",
            "week",
        ],
        optionTitles: [
            "Date",
            "Email",
            "Number",
            "Password",
            "Phone",
            "Search",
            "Text",
            "Time",
            "URL",
            "Week",
        ],
    },
    disabled: {
        type: ControlType.Boolean,
        title: "Disabled",
        defaultValue: defaults.disabled,
    },
    placeholder: {
        type: ControlType.String,
        title: "Placeholder",
        defaultValue: defaults.placeholder,
    },
    value: {
        type: ControlType.String,
        title: "Value",
        defaultValue: defaults.value,
    },
    ...decoratorControls("Search"),
})
