// @framer/framer.base/code/Controls/RadioGroup.tsx
import * as React from "react"
import {
    RadioGroup as BaseRadioGroup,
    RadioGroupProps,
    themes,
} from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { ControlType, addPropertyControls } from "framer"
import { resizeControls, themeControls } from "../propertyControls"
import { centerLayout, resizeLayout, OmitTheme } from "../utils"

type Props = OmitTheme<RadioGroupProps> &
    Partial<{
        theme: string
        resize: string
    }>

const defaults: Props = {
    theme: "light",
    disabled: false,
    resize: "both",
    value: "Option A",
    values: ["Option A", "Option B", "Option C"],
}

export function RadioGroup(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { width, height, resize, theme, ...rest } = props
    const resizeDimensions = resizeLayout(resize, width, height)

    return centerLayout(
        props,
        <BaseRadioGroup
            {...rest}
            theme={themeContext || themes[theme]}
            {...resizeDimensions}
        />
    )
}

RadioGroup.defaultProps = {
    ...defaults,
}

addPropertyControls(RadioGroup, {
    ...themeControls(defaults.theme),
    disabled: {
        type: ControlType.Boolean,
        title: "Disabled",
        defaultValue: defaults.disabled,
    },
    ...resizeControls(defaults.resize),
    value: {
        type: ControlType.String,
        title: "Value",
        defaultValue: defaults.value,
    },
    values: {
        type: ControlType.Array,
        title: "Values",
        defaultValue: defaults.values,
        propertyControl: {
            type: ControlType.String,
            placeholder: "Option…",
        },
    },
})
