// @framer/framer.base/code/Controls/Radio.tsx
import * as React from "react"
import { Radio as BaseRadio, RadioProps, themes } from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { ControlType, addPropertyControls } from "framer"
import { intentControls, themeControls } from "../propertyControls"
import { centerLayout, OmitTheme } from "../utils"

type Props = OmitTheme<RadioProps> &
    Partial<{
        theme: string
    }>

const defaults: Props = {
    theme: "light",
    intent: "primary",
    disabled: false,
    label: "Hello world",
    checked: false,
}

export function Radio(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { theme, ...rest } = props

    return centerLayout(
        props,
        <BaseRadio theme={themeContext || themes[theme]} {...rest} />
    )
}

Radio.defaultProps = {
    ...defaults,
}

addPropertyControls(Radio, {
    ...themeControls(defaults.theme),
    ...intentControls(defaults.intent),
    disabled: {
        type: ControlType.Boolean,
        title: "Disabled",
        defaultValue: defaults.disabled,
    },
    checked: {
        type: ControlType.Boolean,
        title: "Checked",
        defaultValue: defaults.checked,
    },
    label: {
        type: ControlType.String,
        title: "Label",
        defaultValue: defaults.label,
    },
})
