// @framer/framer.base/code/Controls/Slider.tsx
import * as React from "react"
import { Slider as BaseSlider, SliderProps, themes } from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { ControlType, addPropertyControls } from "framer"
import {
    intentControls,
    themeControls,
    resizeControls,
} from "../propertyControls"
import { centerLayout, resizeLayout, OmitTheme } from "../utils"

type Props = OmitTheme<SliderProps> &
    Partial<{
        theme: string
        resize: string
        value: number
        min: number
        max: number
        step: number
    }>

const defaults: Props = {
    theme: "light",
    intent: "primary",
    disabled: false,
    value: 50,
    step: 1,
    min: 0,
    max: 100,
    resize: "none",
}

export function Slider(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { theme, resize, width, height, ...rest } = props
    const resizeDimensions = resizeLayout(resize, width, height)

    return centerLayout(
        props,
        <BaseSlider
            {...rest}
            theme={themeContext || themes[theme]}
            {...resizeDimensions}
        />
    )
}

Slider.defaultProps = {
    ...defaults,
}

addPropertyControls(Slider, {
    ...themeControls(defaults.theme),
    ...intentControls(defaults.intent),
    disabled: {
        type: ControlType.Boolean,
        title: "Disabled",
        defaultValue: defaults.disabled,
    },
    ...resizeControls(defaults.resize),
    value: {
        type: ControlType.Number,
        title: "Value",
        min: defaults.min,
        max: defaults.max,
        step: defaults.step,
        defaultValue: defaults.value,
    },
})
