// @framer/framer.base/code/Controls/Switch.tsx
import * as React from "react"
import { Switch as BaseSwitch, SwitchProps, themes } from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { ControlType, addPropertyControls } from "framer"
import { intentControls, themeControls } from "../propertyControls"
import { centerLayout, OmitTheme } from "../utils"

type Props = OmitTheme<SwitchProps> &
    Partial<{
        theme: string
        label: string
    }>

const defaults: Props = {
    theme: "light",
    intent: "primary",
    disabled: false,
    label: "Hello world",
    checked: false,
}

export function Switch(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { label, theme, ...rest } = props

    return centerLayout(
        props,
        <BaseSwitch theme={themeContext || themes[theme]} {...rest}>
            {label}
        </BaseSwitch>
    )
}

Switch.defaultProps = {
    ...defaults,
}

addPropertyControls(Switch, {
    ...themeControls(defaults.theme),
    ...intentControls(defaults.intent),
    disabled: {
        type: ControlType.Boolean,
        title: "Disabled",
        defaultValue: defaults.disabled,
    },
    checked: {
        type: ControlType.Boolean,
        title: "Checked",
        defaultValue: defaults.checked,
    },
    label: {
        type: ControlType.String,
        title: "Label",
        defaultValue: defaults.label,
    },
})
