// @framer/framer.base/code/Controls/Toggle.tsx
import * as React from "react"
import { Toggle as BaseToggle, ToggleProps, themes } from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { ControlType, addPropertyControls } from "framer"
import {
    intentControls,
    themeControls,
    iconControls,
} from "../propertyControls"
import { centerLayout, OmitTheme } from "../utils"

type Props = OmitTheme<ToggleProps> &
    Partial<{
        theme: string
    }>

const defaults: Props = {
    theme: "light",
    intent: "primary",
    disabled: false,
    icon: "Heart",
    checked: false,
}

export function Toggle(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { theme, ...rest } = props

    return centerLayout(
        props,
        <BaseToggle theme={themeContext || themes[theme]} {...rest} />
    )
}

Toggle.defaultProps = {
    ...defaults,
}

addPropertyControls(Toggle, {
    ...themeControls(defaults.theme),
    ...intentControls(defaults.intent),
    disabled: {
        type: ControlType.Boolean,
        title: "Disabled",
        defaultValue: defaults.disabled,
    },
    checked: {
        type: ControlType.Boolean,
        title: "Checked",
        defaultValue: defaults.checked,
    },
    ...iconControls(defaults.icon),
})
