// @framer/framer.base/code/Theme/Breakpoint.tsx
import * as React from "react"
import { cloneElement, isValidElement, createElement } from "react"
import { themes } from "../../../base"
import { placeholderState } from "../utils/"
import { breakpointControls } from "../propertyControls"
import { Frame, addPropertyControls, RenderTarget } from "framer"

const breakpoints = {
    names: Object.keys(themes.light.breakpoint),
    values: Object.keys(themes.light.breakpoint).map(key =>
        parseInt(themes.light.breakpoint[key])
    ),
}

interface Props {
    width?: number
    height?: number
}

export function Breakpoint(props: Props) {
    const { width } = props
    const smallerBreakpoints = breakpoints.values.filter((breakpoint, index) =>
        !index ? true : breakpoint < width
    )
    const layouts = breakpoints.names
        .filter(
            (_, index) =>
                smallerBreakpoints.indexOf(breakpoints.values[index]) >= 0
        )
        .map(breakpoint => props[breakpoint][0])
        .filter(layout => layout)

    const layout = layouts.length ? layouts[layouts.length - 1] : <></>

    return layouts.length === 0 &&
        (RenderTarget.current() === RenderTarget.canvas ||
            RenderTarget.current() === RenderTarget.thumbnail) ? (
        createElement(placeholderState, {
            theme: themes.light,
            title: "No breakpoints",
            label: "Add breakpoints by connecting them on the Canvas",
        })
    ) : (
        <Frame background="transparent" size="100%">
            {isValidElement(layout) &&
                cloneElement(layout, {
                    width: "100%",
                    height: "100%",
                })}
        </Frame>
    )
}

addPropertyControls(Breakpoint, {
    ...breakpointControls(breakpoints),
})
