// @framer/framer.base/code/Theme/Theme.tsx
import * as React from "react"
import { isValidElement, cloneElement, createElement } from "react"
import { themes, capitalize } from "../../../base"
import { Frame, ControlType, RenderTarget, addPropertyControls } from "framer"
import { useMutationObserver } from "../hooks"
import { placeholderState } from "../utils"
import { ThemeProvider } from "styled-components"

interface Props {
    width?: number
    height?: number
    children?: any
    background?: string
    theme: string
}

const defaults: Props = {
    theme: "light",
}

const colors = Object.keys(themes.light.color)
const defaultThemeOptions = Object.keys(themes)
const themeOptions = ["automatic", ...defaultThemeOptions]

export function Theme(props: Props = defaults) {
    const automaticTheme =
        useMutationObserver(
            document.body,
            {
                attributes: true,
                attributeFilter: ["class"],
            },
            ([mutation]) => mutation.target.className
        ) || defaults.theme
    const { children, background, theme } = props
    const selectedTheme =
        themes[defaultThemeOptions.indexOf(theme) >= 0 ? theme : automaticTheme]
    const [child] = children

    return !child &&
        (RenderTarget.current() === RenderTarget.canvas ||
            RenderTarget.current() === RenderTarget.thumbnail) ? (
        createElement(placeholderState, {
            theme: themes.light,
            title: "No content",
            label: "Add content by connecting it on the Canvas",
        })
    ) : (
        <ThemeProvider theme={selectedTheme}>
            {child && (
                <Frame
                    background={
                        selectedTheme.color[background] || "transparent"
                    }
                    size="100%"
                >
                    {isValidElement(child) &&
                        cloneElement(child, {
                            width: "100%",
                            height: "100%",
                        })}
                </Frame>
            )}
        </ThemeProvider>
    )
}

Theme.defaultProps = {
    ...defaults,
}

addPropertyControls(Theme, {
    theme: {
        type: ControlType.Enum,
        title: "Theme",
        defaultValue: defaults.theme,
        options: themeOptions,
        optionTitles: themeOptions.map(themeOption => capitalize(themeOption)),
    },
    background: {
        type: ControlType.Enum,
        title: "Background",
        defaultValue: "background",
        options: colors,
        optionTitles: colors.map(
            color => color.charAt(0).toUpperCase() + color.slice(1)
        ),
    },
    children: {
        type: ControlType.ComponentInstance,
        title: "Content",
    },
})
