// @framer/framer.base/code/Typography/Heading.tsx
import * as React from "react"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { Heading as BaseHeading, HeadingProps, themes } from "../../../base"
import { ControlType, addPropertyControls } from "framer"
import { themeControls } from "../propertyControls"
import { centerLayout, resizeLayout, OmitTheme } from "../utils"

type Props = OmitTheme<HeadingProps> &
    Partial<{
        theme: string
        text: string | any
        resize: string
    }>

const defaults: Props = {
    theme: "light",
    text: "Hello world",
    level: 2,
    resize: "none",
}

export function Heading(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { width, height, level, resize, text, theme, ...rest } = props
    const resizeDimensions = resizeLayout(resize, width, height)

    return centerLayout(
        props,
        <BaseHeading
            {...rest}
            level={Number(level) as 1 | 2 | 3 | 4 | 5 | 6}
            theme={themeContext || themes[theme]}
            {...resizeDimensions}
        >
            {text}
        </BaseHeading>
    )
}

Heading.defaultProps = {
    ...defaults,
}

addPropertyControls(Heading, {
    ...themeControls(defaults.theme),
    level: {
        type: ControlType.Enum,
        title: "Level",
        defaultValue: "2",
        options: ["1", "2", "3", "4", "5", "6"],
        optionTitles: ["H1", "H2", "H3", "H4", "H5", "H6"],
    },
    align: {
        type: ControlType.Enum,
        title: "Align",
        defaultValue: "left",
        options: ["left", "center", "right", "justify"],
        optionTitles: ["Left", "Center", "Right", "Justify"],
    },
    text: {
        type: ControlType.String,
        title: "Text",
        defaultValue: defaults.text,
    },
})
