// @framer/framer.base/code/Typography/Text.tsx
import * as React from "react"
import { Text as BaseText, TextProps, themes } from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { ControlType, addPropertyControls } from "framer"
import { themeControls } from "../propertyControls"
import { centerLayout, resizeLayout, OmitTheme } from "../utils"

type Props = OmitTheme<TextProps> &
    Partial<{
        theme: string
        text: string | any
        resize: string
    }>

const defaults: Props = {
    theme: "light",
    text: "Hello world",
    small: false,
    code: false,
    resize: "none",
}

export function Text(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { width, height, resize, text, theme, ...rest } = props
    const resizeDimensions = resizeLayout(resize, width, height)

    return centerLayout(
        props,
        <BaseText
            {...rest}
            theme={themeContext || themes[theme]}
            {...resizeDimensions}
        >
            {text}
        </BaseText>
    )
}

Text.defaultProps = {
    ...defaults,
}

addPropertyControls(Text, {
    ...themeControls(defaults.theme),
    small: {
        type: ControlType.Boolean,
        title: "Small",
        defaultValue: defaults.small,
    },
    code: {
        type: ControlType.Boolean,
        title: "Font",
        defaultValue: defaults.code,
        disabledTitle: "Sans",
        enabledTitle: "Mono",
    },
    align: {
        type: ControlType.Enum,
        title: "Align",
        defaultValue: "left",
        options: ["left", "center", "right", "justify"],
        optionTitles: ["Left", "Center", "Right", "Justify"],
    },
    text: {
        type: ControlType.String,
        title: "Text",
        defaultValue: defaults.text,
    },
})
