// @framer/framer.base/code/Miscellaneous/Image.tsx
import * as React from "react"
import { useContext, createElement } from "react"
import { ThemeContext } from "styled-components"
import { themes } from "../../../base"
import { intentControls, themeControls } from "../propertyControls"
import { placeholderState } from "../utils"
import {
    Frame,
    FrameProps,
    addPropertyControls,
    ControlType,
    RenderTarget,
} from "framer"

type Props = Partial<
    FrameProps & {
        theme: string
        intent: string
        image: string
        overlay: boolean
        color: string
    }
>

const defaults: Props = {
    theme: "light",
    intent: "primary",
    overlay: false,
}

export function Image(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { theme, intent, image, overlay, color, ...rest } = props
    const dynamicTheme = themeContext || themes[theme]

    return (
        <Frame background="transparent" size="100%" {...rest}>
            {!image &&
            (RenderTarget.current() === RenderTarget.canvas ||
                RenderTarget.current() === RenderTarget.thumbnail) ? (
                createElement(placeholderState, {
                    theme: dynamicTheme,
                    title: "No image",
                    label: "Choose an image in the Properties panel",
                })
            ) : (
                <>
                    <Frame
                        image={image}
                        size="100%"
                        style={{
                            backgroundColor: "transparent",
                            backgroundSize: "cover",
                            backgroundPosition: "center",
                            WebkitFilter:
                                overlay && "contrast(77%) grayscale(100%)",
                        }}
                    />
                    {overlay && (
                        <Frame
                            background={color || dynamicTheme.color[intent]}
                            opacity={0.5}
                            size="100%"
                            style={{
                                mixBlendMode: "overlay",
                            }}
                        />
                    )}
                </>
            )}
        </Frame>
    )
}

Image.defaultProps = {
    ...defaults,
}

addPropertyControls(Image, {
    ...themeControls(defaults.theme, true, "overlay"),
    ...intentControls(defaults.intent, true, "overlay"),
    image: {
        type: ControlType.Image,
        title: "Image",
    },
    overlay: {
        type: ControlType.Boolean,
        title: "Overlay",
        enabledTitle: "Visible",
        disabledTitle: "Hidden",
        defaultValue: defaults.overlay,
    },
})
