// @framer/framer.base/code/Miscellaneous/Progress.tsx
import * as React from "react"
import { Progress as BaseProgress, ProgressProps, themes } from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { ControlType, addPropertyControls } from "framer"
import {
    intentControls,
    themeControls,
    resizeControls,
} from "../propertyControls"
import { centerLayout, resizeLayout, OmitTheme } from "../utils"

type Props = OmitTheme<ProgressProps> &
    Partial<{
        theme: string
        value: number
        resize: string
    }>

const defaults: Props = {
    width: 320,
    height: 6,
    theme: "light",
    intent: "primary",
    value: 50,
    max: 100,
    resize: "none",
}

export function Progress(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { theme, value, max, resize, width, height, ...rest } = props
    const resizeDimensions = resizeLayout(resize, width, height)

    return centerLayout(
        props,
        <BaseProgress
            {...rest}
            max={Number(max)}
            theme={themeContext || themes[theme]}
            value={Number(value)}
            {...resizeDimensions}
        />
    )
}

Progress.defaultProps = {
    ...defaults,
}

addPropertyControls(Progress, {
    ...themeControls(defaults.theme),
    ...intentControls(defaults.intent),
    value: {
        type: ControlType.Number,
        title: "Value",
        min: 0,
        max: defaults.max,
        defaultValue: defaults.value,
    },
    ...resizeControls(defaults.resize),
})
