// @framer/framer.base/code/Miscellaneous/Tooltip.tsx
import * as React from "react"
import { Tooltip as BaseTooltip, TooltipProps, themes } from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { ControlType, addPropertyControls } from "framer"
import { intentControls, themeControls } from "../propertyControls"
import { centerLayout, OmitTheme } from "../utils"

type Props = OmitTheme<TooltipProps> &
    Partial<{
        theme: string
        text: string
    }>

const defaults: Props = {
    theme: "light",
    intent: "primary",
    text: "Hello world",
    tooltip: "Hello world",
    direction: "top",
}

export function Tooltip(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { text, theme, ...rest } = props

    return centerLayout(
        props,
        <BaseTooltip theme={themeContext || themes[theme]} {...rest}>
            {text}
        </BaseTooltip>
    )
}

Tooltip.defaultProps = {
    ...defaults,
}

addPropertyControls(Tooltip, {
    ...themeControls(defaults.theme),
    ...intentControls(defaults.intent),
    direction: {
        type: ControlType.SegmentedEnum,
        title: "Direction",
        defaultValue: defaults.direction,
        options: ["top", "right", "bottom", "left"],
        optionTitles: ["↑", "→", "↓", "←"],
    },
    text: {
        type: ControlType.String,
        title: "Text",
        defaultValue: defaults.text,
    },
    tooltip: {
        type: ControlType.String,
        title: "Tooltip",
        defaultValue: defaults.tooltip,
    },
})
