// @framer/framer.base/code/Overrides.tsx
import { Override } from "framer"

export function onButtonClick(): Override {
    return {
        onClick: () => {
            console.log("Hello world")
        },
    }
}

export function setDecorators(): Override {
    return {
        startDecorator: "ZoomOut",
        endDecorator: "ZoomIn",
    }
}

export function getSliderValue(): Override {
    return {
        onValueChange: value => {
            console.log(value)
        },
    }
}

export function setSliderValue(): Override {
    return {
        value: 0,
    }
}

export function getInputValue(): Override {
    return {
        onValueChange: value => {
            console.log(value)
        },
    }
}

export function setInputValue(): Override {
    return {
        value: "Hello world",
    }
}

export function getSelectValue(): Override {
    return {
        onValueChange: value => {
            console.log(value)
        },
    }
}

export function setSelectValue(): Override {
    return {
        value: "Hello world",
    }
}

export function setProgressValue(): Override {
    return {
        value: 1,
    }
}

export function setBackgroundColor(): Override {
    return {
        backgroundColor: "#09f",
    }
}

export function setColor(): Override {
    return {
        color: "#09f",
    }
}

export function setPattern(): Override {
    return {
        pattern: "secondary",
    }
}
