// @framer/framer.base/code/Layout/Base.tsx
import * as React from "react"
import { useContext, cloneElement, isValidElement } from "react"
import { ThemeContext } from "styled-components"
import { themes, Theme } from "../../../base"
import { themeControls } from "../propertyControls"
import { Frame, FrameProps, addPropertyControls, ControlType } from "framer"

type Props = Partial<
    Omit<FrameProps, "shadow" | "radius"> & {
        theme: string
        color: string
        shadow: number
        radius: number
        children: any
    }
>

const defaults: Props = {
    theme: "light",
    color: "background",
    shadow: 0,
    radius: 0,
}

const elevationShadow = (shadow: number, theme: Theme): string => {
    if (shadow === 1) {
        return `${theme.shadow.smaller}`
    } else if (shadow === 2) {
        return `${theme.shadow.small}, ${theme.shadow.medium}`
    } else if (shadow === 3) {
        return `${theme.shadow.medium}, ${theme.shadow.large}`
    } else if (shadow === 4) {
        return `${theme.shadow.large}, ${theme.shadow.larger}`
    } else {
        return ""
    }
}

const colors = Object.keys(themes.light.color)

export function Base(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const { theme, color, shadow, radius, children, ...rest } = props
    const dynamicTheme = themeContext || themes[theme]
    const [child] = children

    return (
        <Frame
            {...rest}
            background={dynamicTheme.color[color] || dynamicTheme.color.white}
            radius={dynamicTheme.radius[radius] || 0}
            shadow={elevationShadow(shadow, dynamicTheme) || ""}
            size="100%"
        >
            {isValidElement(child) &&
                cloneElement(child, {
                    width: "100%",
                    height: "100%",
                })}
        </Frame>
    )
}

Base.defaultProps = {
    ...defaults,
}

addPropertyControls(Base, {
    ...themeControls(defaults.theme),
    color: {
        type: ControlType.Enum,
        title: "Color",
        defaultValue: "background",
        options: colors,
        optionTitles: colors.map(
            color => color.charAt(0).toUpperCase() + color.slice(1)
        ),
    },
    shadow: {
        type: ControlType.Number,
        title: "Shadow",
        defaultValue: defaults.shadow,
        min: 0,
        max: 4,
        displayStepper: true,
    },
    radius: {
        type: ControlType.Number,
        title: "Radius",
        min: 0,
        max: themes[defaults.theme].radius.length - 1,
        displayStepper: true,
    },
    children: {
        type: ControlType.ComponentInstance,
        title: "Content",
    },
})
