// @framer/framer.base/code/Layout/Card.tsx
import * as React from "react"
import {
    Card as BaseCard,
    CardProps,
    Heading,
    Text,
    Toggle,
    themes,
    parseColor,
} from "../../../base"
import { useContext } from "react"
import { ThemeContext } from "styled-components"
import { Frame, Stack, ControlType, addPropertyControls } from "framer"
import { Intent } from "../../../base/modifiers"
import {
    intentControls,
    themeControls,
    iconControls,
} from "../propertyControls"
import { centerLayout, OmitTheme } from "../utils"
import { Image } from "../Miscellaneous/Image"

type Props = OmitTheme<CardProps> &
    Partial<{
        theme: string
        intent: Intent
        image: string
        title: string
        body: string
        footer: string
        withToggle: boolean
        iconToggle: string
        toggle: string
        overlay: boolean
    }>

const defaults: Props = {
    theme: "light",
    intent: "primary",
    image: "",
    overlay: false,
    title: "",
    body: "",
    footer: "",
    toggle: "",
    withToggle: false,
    iconToggle: "Heart",
    shadow: 2,
}

export function Card(props: Props = defaults) {
    const themeContext = useContext(ThemeContext)
    const {
        width,
        height,
        color,
        image,
        overlay,
        title,
        body,
        footer,
        theme,
        intent,
        toggle,
        withToggle,
        iconToggle,
        ...rest
    } = props
    const dynamicToggle = withToggle ? iconToggle : toggle
    const dynamicTheme = themeContext || themes[theme]
    const toggleComponent = (intent: Intent) =>
        dynamicToggle && (
            <Toggle
                color={image && parseColor(dynamicTheme.color.white)}
                icon={dynamicToggle}
                intent={intent}
                style={{
                    position: "absolute",
                    top: "var(--card-padding)",
                    right: "var(--card-padding)",
                }}
                theme={dynamicTheme}
            />
        )

    return centerLayout(
        props,
        <BaseCard
            color={color}
            height={height}
            width={width}
            {...rest}
            padding={false}
            theme={dynamicTheme}
        >
            <Stack
                background="transparent"
                direction="vertical"
                gap={0}
                size="100%"
            >
                {image && (
                    <Frame
                        background="transparent"
                        height="1fr"
                        overflow="hidden"
                        width="1fr"
                    >
                        <Image
                            image={image}
                            intent={intent}
                            overlay={overlay}
                        />
                        {toggleComponent("text")}
                    </Frame>
                )}
                {(title || body || ((withToggle || toggle) && !image)) && (
                    <Frame
                        background="transparent"
                        height="auto"
                        style={{
                            padding: "var(--card-padding)",
                        }}
                        width="100%"
                    >
                        {title && (
                            <Heading
                                color={color}
                                level={4}
                                style={{
                                    marginBottom:
                                        (body || footer) &&
                                        "calc(var(--card-padding) / 2)",
                                }}
                                theme={themeContext || themes[theme]}
                            >
                                {title}
                            </Heading>
                        )}
                        {body && (
                            <Text
                                color={color}
                                theme={themeContext || themes[theme]}
                            >
                                {body}
                            </Text>
                        )}
                        {footer && (
                            <Text
                                color={color}
                                small
                                style={{
                                    opacity: 0.5,
                                    marginTop:
                                        (body || title) &&
                                        "calc(var(--card-padding) / 2)",
                                }}
                                theme={themeContext || themes[theme]}
                            >
                                {footer}
                            </Text>
                        )}
                        {!image && toggleComponent(intent)}
                    </Frame>
                )}
            </Stack>
        </BaseCard>
    )
}

Card.defaultProps = {
    ...defaults,
}

addPropertyControls(Card, {
    ...themeControls(defaults.theme),
    ...intentControls(defaults.intent),
    image: {
        type: ControlType.Image,
        title: "Image",
    },
    overlay: {
        type: ControlType.Boolean,
        title: "Overlay",
        enabledTitle: "Visible",
        disabledTitle: "Hidden",
        defaultValue: defaults.overlay,
    },
    withToggle: {
        type: ControlType.Boolean,
        title: "Toggle",
        defaultValue: false,
        enabledTitle: "Icon",
        disabledTitle: "None",
    },
    ...iconControls(defaults.iconToggle, false, "withToggle", "iconToggle"),
    title: {
        type: ControlType.String,
        title: "Title",
        defaultValue: defaults.title,
    },
    body: {
        type: ControlType.String,
        title: "Body",
        defaultValue: defaults.body,
    },
    footer: {
        type: ControlType.String,
        title: "Footer",
        defaultValue: defaults.footer,
    },
    shadow: {
        type: ControlType.Number,
        title: "Shadow",
        defaultValue: defaults.shadow,
        min: 0,
        max: 4,
        displayStepper: true,
    },
})
