// @framer/framer.base/code/Layout/List.tsx
import * as React from "react"
import { createElement } from "react"
import { themes } from "../../../base"
import { placeholderState } from "../utils"
import {
    ScrollProps,
    Stack,
    Scroll,
    addPropertyControls,
    ControlType,
    RenderTarget,
} from "framer"

type Props = Partial<
    ScrollProps & {
        gap: number
        scroll: boolean
        padding: number
        paddingTop: number
        paddingLeft: number
        paddingRight: number
        paddingBottom: number
        paddingPerSide: boolean
    }
> & {
    content: any[]
}

const defaults: Props = {
    content: [],
    gap: 8,
    scroll: true,
    padding: 0,
}

export function List(props: Props = defaults) {
    const {
        content,
        gap,
        scroll,
        paddingPerSide,
        padding,
        paddingTop,
        paddingLeft,
        paddingRight,
        paddingBottom,
        ...rest
    } = props

    const paddings = paddingPerSide
        ? {
              paddingRight,
              paddingBottom,
              paddingLeft,
              paddingTop,
          }
        : paddingPerSide === undefined
        ? {
              padding,
              paddingRight: paddingRight || padding,
              paddingBottom: paddingBottom || padding,
              paddingLeft: paddingLeft || padding,
              paddingTop: paddingTop || padding,
          }
        : {
              padding,
          }

    return content.length === 0 &&
        (RenderTarget.current() === RenderTarget.canvas ||
            RenderTarget.current() === RenderTarget.thumbnail) ? (
        createElement(placeholderState, {
            theme: themes.light,
            title: "No items",
            label: "Add items by connecting them on the Canvas",
        })
    ) : (
        <Scroll {...rest} dragEnabled={scroll}>
            <Stack
                background="transparent"
                direction="vertical"
                gap={gap}
                width="100%"
                {...paddings}
            >
                {content}
            </Stack>
        </Scroll>
    )
}

List.defaultProps = {
    ...defaults,
}

addPropertyControls(List, {
    gap: {
        type: ControlType.Number,
        title: "Gap",
        displayStepper: true,
        defaultValue: defaults.gap,
        min: 0,
    },
    padding: {
        type: ControlType.FusedNumber,
        title: "Padding",
        toggleKey: "paddingPerSide",
        toggleTitles: ["All Sides", "Per Side"],
        valueKeys: [
            "paddingTop",
            "paddingRight",
            "paddingBottom",
            "paddingLeft",
        ],
        valueLabels: ["T", "R", "B", "L"],
        defaultValue: defaults.padding,
        min: 0,
    },
    content: {
        title: "Content",
        type: ControlType.Array,
        propertyControl: {
            type: ControlType.ComponentInstance,
        },
        defaultValue: defaults.content,
    },
})
